### DISCUSSION TITLE
Fast Api Multi Folder

### DISCUSSION THREAD

**Discussion by pierresigwalt** [litestar-org/advanced-alchemy#232] [Q&A]
So, I have followed the FastApi Example, everything works when everything is in one file. However I am having hard time when trying to put some order in the organisation (Circular dependency) and can't see how to avoid this (I kind of follow litestart organisation because I like the approach):

My Current Setup is the following:

app/
├── main.py
├── domain
│   ├── dependencies.py
├── server
│   ├── dependencies.py

* **main.py**
```python
idp_app = FastAPI(
    swagger_ui_parameters={
        "syntaxHighlight.theme": "obsidian"
    },
    lifespan=lifespan,
    title=openapi_info["title"],
    version=openapi_info["version"],
    description=openapi_info["description"],
    contact=openapi_info["contact"],
    license_info=openapi_info["license_info"],
)

for router in app_routers:
    idp_app.include_router(router)


alchemy = StarletteAdvancedAlchemy(config=db_config, app=idp_app)

````

* **domain/dependencies.py**
```python
async def provide_user_service(
    db_session: Annotated[AsyncSession, Depends(provide_db_session)],
) -> AsyncGenerator[UserService, None]:
    """This provides the default Authors repository."""
    async with UserService.new(
        session=db_session,
    ) as service:
        yield service

```

* **server/dependencies.py**
```python
async def provide_db_session(request: Request) -> AsyncSession:
    """Provide a DB session."""
    return alchemy.get_session(request)
```

However as main.py alchemy variable need the app (FastApi) I can't really move it out from the main.py from my knowledge off course, bringing the issue of the dependency.

So the question is:
How can I create alchemy->provide_db_session outside of the main.py and be able to include the app to StarletteAdvancedAlchemy

Many thank for the help, hopefully it's clear.

**Comment by nvanmeurs** [comment]
@pierresigwalt Did you ever figure it out?

**Comment by Priestch** [comment]
I think you can initialize the `alchemy` object with the default app, which is None by default in a individual module, then import and  call it's `init_app` in main.py, the import `alchemy` statement in other modules should work.

https://github.com/litestar-org/advanced-alchemy/blob/95492689c24886c873ae8cfb0349c35d44a2be5d/advanced_alchemy/extensions/starlette.py#L112-L132

**Reply by Alc-Alc** [reply]
@pierresigwalt, @nvanmeurs this is indeed how you can do this.

pseudo code

domain/dependencies.py
```
sqlalchemy_config = SQLAlchemyAsyncConfig(
    connection_string="sqlite+aiosqlite:///test.sqlite",
    session_config=AsyncSessionConfig(expire_on_commit=False),
    create_all=True,
)

alchemy = AdvancedAlchemy(config=sqlalchemy_config)

DatabaseSession = Annotated[AsyncSession, Depends(alchemy.provide_session())]

async def provide_authors_service(db_session: DatabaseSession) -> AsyncGenerator[AuthorService, None]:
    """This provides the default Authors repository."""
    async with AuthorService.new(session=db_session) as service:
        yield service
```

main.py
```
from domain.dependencies import alchemy, provide_authors_service ,...

app = FastAPI()

alchemy.init_app(app)
```
